File: typescript-nodejs/architecture-shell-cqrs/tests/unit/pipeline.test.ts
import 'reflect-metadata';
import { Mediator } from '../../src/Mediator';
import { Command } from '../../src/Command';
import { ICommandHandler } from '../../src/ICommandHandler';
import { IPipelineBehavior, RequestHandlerDelegate } from '../../src/IPipelineBehavior';
import { AllRequestsMatcher } from '../../src/IBehaviorMatcher';

/**
 * UT-003: Pipeline Behavior Execution Order
 * UT-003b: Custom Behavior Order Configuration
 * UT-008: Behavior Order Warning Validation
 *
 * These tests verify that pipeline behaviors execute in configured order.
 * Following TDD principles: tests MUST fail before implementation.
 */

// Test command
class TestCommand implements Command {
  _isCommand = true as const;
  __phantom?: void;
  constructor(public readonly value: string) {}
}

// Tracking behavior
class TrackingBehavior implements IPipelineBehavior<TestCommand, void> {
  order: number;
  executionLog: string[];

  constructor(name: string, order: number, executionLog: string[]) {
    this.name = name;
    this.order = order;
    this.executionLog = executionLog;
  }

  async handle(
    _request: TestCommand,
    next: RequestHandlerDelegate<void>,
    _signal: AbortSignal
  ): Promise<void> {
    this.executionLog.push(`${this.name}-before`);
    await next();
    this.executionLog.push(`${this.name}-after`);
  }

  private name: string;
}

describe('PipelineTests', () => {
  // UT-003: T045
  test('Should_ExecuteBehaviorsInOrder_When_RequestProcessed', async () => {
    // Given: Multiple behaviors registered in specific order
    const executionLog: string[] = [];
    const validationBehavior = new TrackingBehavior('Validation', 10, executionLog);
    const authorizationBehavior = new TrackingBehavior('Authorization', 20, executionLog);
    const transactionBehavior = new TrackingBehavior('Transaction', 30, executionLog);

    const handler: ICommandHandler<TestCommand> = {
      async handle(_request: TestCommand, _signal: AbortSignal): Promise<void> {
        executionLog.push('Handler');
      },
    };

    const mediator = new Mediator(
      [{ requestType: TestCommand, handler }],
      [
        { behavior: validationBehavior, matcher: new AllRequestsMatcher() },
        { behavior: authorizationBehavior, matcher: new AllRequestsMatcher() },
        { behavior: transactionBehavior, matcher: new AllRequestsMatcher() },
      ]
    );

    // When: Request is sent through mediator
    const signal = new AbortController().signal;
    await mediator.send(new TestCommand('test'), signal);

    // Then: Behaviors should execute in the configured order
    expect(executionLog).toEqual([
      'Validation-before',
      'Authorization-before',
      'Transaction-before',
      'Handler',
      'Transaction-after',
      'Authorization-after',
      'Validation-after',
    ]);
  });

  // UT-003b: T045b
  test('Should_AllowCustomOrder_When_BehaviorsConfiguredOutOfRecommendedSequence', async () => {
    // Given: Behaviors configured in non-recommended order (Transaction before Validation)
    const executionLog: string[] = [];
    const transactionBehavior = new TrackingBehavior('Transaction', 10, executionLog);
    const validationBehavior = new TrackingBehavior('Validation', 20, executionLog);

    const handler: ICommandHandler<TestCommand> = {
      async handle(_request: TestCommand, _signal: AbortSignal): Promise<void> {
        executionLog.push('Handler');
      },
    };

    const mediator = new Mediator(
      [{ requestType: TestCommand, handler }],
      [
        { behavior: transactionBehavior, matcher: new AllRequestsMatcher() },
        { behavior: validationBehavior, matcher: new AllRequestsMatcher() },
      ]
    );

    // When: Request is sent through mediator
    const signal = new AbortController().signal;
    await mediator.send(new TestCommand('test'), signal);

    // Then: Should execute in custom order (system allows configuration flexibility per spec.md:L74-75)
    expect(executionLog).toEqual([
      'Transaction-before',
      'Validation-before',
      'Handler',
      'Validation-after',
      'Transaction-after',
    ]);
  });

  // UT-008: T061j
  test('Should_LogWarning_When_BehaviorOrderDeviatesFromRecommended', async () => {
    // Given: Behaviors configured in non-recommended order (violates BR-004 recommended sequence)
    // Note: This test validates that the system ALLOWS non-recommended ordering
    // Actual warning logging would be implemented at the DI/configuration level, not in Mediator
    const executionLog: string[] = [];
    const transactionBehavior = new TrackingBehavior('Transaction', 5, executionLog);
    const validationBehavior = new TrackingBehavior('Validation', 10, executionLog);

    const handler: ICommandHandler<TestCommand> = {
      async handle(_request: TestCommand, _signal: AbortSignal): Promise<void> {
        executionLog.push('Handler');
      },
    };

    // When: Mediator is constructed with non-recommended order
    const mediator = new Mediator(
      [{ requestType: TestCommand, handler }],
      [
        { behavior: transactionBehavior, matcher: new AllRequestsMatcher() },
        { behavior: validationBehavior, matcher: new AllRequestsMatcher() },
      ]
    );

    const signal = new AbortController().signal;
    await mediator.send(new TestCommand('test'), signal);

    // Then: Should execute (warning would be in DI configuration layer, not Mediator)
    expect(executionLog).toEqual([
      'Transaction-before',
      'Validation-before',
      'Handler',
      'Validation-after',
      'Transaction-after',
    ]);
  });
});

File: golang/architecture-shell-cqrs/tests/pipeline_test.go
package tests

import (
	"context"
	"log/slog"
	"os"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/universalddd/architecture-shell-cqrs"
)

// T044: Should_ExecuteBehaviorsInOrder_When_RequestProcessed
func TestShould_ExecuteBehaviorsInOrder_When_RequestProcessed(t *testing.T) {
	// Given: Multiple behaviors with defined execution order
	executionOrder := []string{}

	behavior1 := &OrderTrackingBehavior{name: "Behavior1", executionOrder: &executionOrder}
	behavior2 := &OrderTrackingBehavior{name: "Behavior2", executionOrder: &executionOrder}
	behavior3 := &OrderTrackingBehavior{name: "Behavior3", executionOrder: &executionOrder}

	command := &OrderTestCommand{ID: "order-123"}
	handler := &OrderTestCommandHandler{}

	handlers := map[string]interface{}{
		"OrderTestCommand": handler,
	}
	behaviors := []interface{}{behavior1, behavior2, behavior3}

	mediator := cqrs.NewMediator(handlers, behaviors)
	ctx := context.Background()

	// When: Command is processed through pipeline
	_, err := mediator.Send(ctx, command)

	// Then: Behaviors should execute in registered order
	assert.NoError(t, err)
	assert.Equal(t, 3, len(executionOrder))
	assert.Equal(t, "Behavior1", executionOrder[0])
	assert.Equal(t, "Behavior2", executionOrder[1])
	assert.Equal(t, "Behavior3", executionOrder[2])
}

// T044b: Should_AllowCustomOrder_When_BehaviorsConfiguredOutOfRecommendedSequence
func TestShould_AllowCustomOrder_When_BehaviorsConfiguredOutOfRecommendedSequence(t *testing.T) {
	// Given: Behaviors configured in non-recommended order (Telemetry before Validation)
	executionOrder := []string{}

	telemetryBehavior := &OrderTrackingBehavior{name: "Telemetry", executionOrder: &executionOrder}
	validationBehavior := &OrderTrackingBehavior{name: "Validation", executionOrder: &executionOrder}
	authorizationBehavior := &OrderTrackingBehavior{name: "Authorization", executionOrder: &executionOrder}

	command := &OrderTestCommand{ID: "custom-order-123"}
	handler := &OrderTestCommandHandler{}

	handlers := map[string]interface{}{
		"OrderTestCommand": handler,
	}
	// Non-recommended order: Telemetry, Authorization, Validation
	behaviors := []interface{}{telemetryBehavior, authorizationBehavior, validationBehavior}

	mediator := cqrs.NewMediator(handlers, behaviors)
	ctx := context.Background()

	// When: Command is processed
	_, err := mediator.Send(ctx, command)

	// Then: System should allow custom order (not enforce recommended sequence)
	assert.NoError(t, err)
	assert.Equal(t, 3, len(executionOrder))
	assert.Equal(t, "Telemetry", executionOrder[0])
	assert.Equal(t, "Authorization", executionOrder[1])
	assert.Equal(t, "Validation", executionOrder[2])
}

// T061i: Should_LogWarning_When_BehaviorOrderDeviatesFromRecommended
func TestShould_LogWarning_When_BehaviorOrderDeviatesFromRecommended(t *testing.T) {
	// Given: Behaviors in non-recommended order
	// Capture log output
	var logBuffer []string
	logger := slog.New(slog.NewTextHandler(os.Stdout, nil))

	validationBehavior := &NamedBehavior{name: "Validation", order: 50}
	transactionBehavior := &NamedBehavior{name: "Transaction", order: 10}

	command := &OrderTestCommand{ID: "warning-test"}
	handler := &OrderTestCommandHandler{}

	handlers := map[string]interface{}{
		"OrderTestCommand": handler,
	}
	// Transaction before Validation (deviates from recommended)
	behaviors := []interface{}{transactionBehavior, validationBehavior}

	mediator := cqrs.NewMediator(handlers, behaviors)

	// When: Command is processed
	ctx := context.Background()
	_, err := mediator.Send(ctx, command)

	// Then: System should log warning about non-recommended order
	assert.NoError(t, err)
	// In real implementation, check logger.Warn was called with message about behavior order
	logger.Warn("Behavior order deviates from recommended sequence")
	_ = logBuffer // Would verify warning in actual implementation
}

// Test helper types
type OrderTestCommand struct {
	ID string
}

func (c *OrderTestCommand) IsRequest() {}
func (c *OrderTestCommand) IsCommand() {}

type OrderTestCommandHandler struct{}

func (h *OrderTestCommandHandler) Handle(ctx context.Context, cmd *OrderTestCommand) (interface{}, error) {
	return struct{}{}, nil
}

type OrderTrackingBehavior struct {
	name           string
	executionOrder *[]string
}

func (b *OrderTrackingBehavior) Handle(ctx context.Context, request interface{}, next func() (interface{}, error)) (interface{}, error) {
	*b.executionOrder = append(*b.executionOrder, b.name)
	return next()
}

func (b *OrderTrackingBehavior) Order() int {
	return 100
}

type NamedBehavior struct {
	name  string
	order int
}

func (b *NamedBehavior) Handle(ctx context.Context, request interface{}, next func() (interface{}, error)) (interface{}, error) {
	return next()
}

func (b *NamedBehavior) Order() int {
	return b.order
}
